### PR TITLE
Change definition of RepairTask

### PR DESCRIPTION
**Pull request by emolsson**
* Create RepairGroup as a task to TableRepairJob
 - Make RepairTask an operation of RepairGroup
* Move execution/locking to tasks (RepairGroup)
* Create RepairStateSnapshot used by the task (RepairGroup)
 - Snapshot view of what needs to be repaired, what locks need to be taken, etc.

### CODE REVIEW

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/RepairGroup.java [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+
+    @Override
+    public LockFactory.DistributedLock getLock(LockFactory lockFactory) throws LockException
+    {
+        Collection<String> dataCenters = myRepairStateSnapshot.getDatacentersForRepair();
+        for (String dc : dataCenters)
+        {
+            if (!lockFactory.sufficientNodesForLocking(dc, getResource(dc, 1)))
+            {
+                throw new LockException("Data center " + dc + " not lockable. Repair will be retried later");
```
[the remark this answers is not in the document]
Comment (emolsson) [reply core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/RepairGroup.java]: Done

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/RepairGroup.java [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+     *
+     * @param lockFactory
+     *            The lock factory
+     * @param dataCenter
+     *            The data center to check
+     * @return The total number of locks acquired for this table
+     */
+    private int getTableRepairLockCount(LockFactory lockFactory, String dataCenter)
+    {
+        int ret = 0;
```
[the remark this answers is not in the document]
Comment (emolsson) [reply core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/RepairGroup.java]: Logic has been simplified somewhat in later commits, so this method will be obsolete. Is it ok if I leave it as is for now?

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/RepairState.java [hunk 3]
```diff
@@ -14,71 +14,20 @@
...
 
-import java.util.Collection;
-import java.util.Map;
-import java.util.Set;
-
-import com.datastax.driver.core.Host;
-import com.ericsson.bss.cassandra.ecchronos.core.utils.LongTokenRange;
-
 /**
  * Interface used by TableRepairJob to maintain the current repair state.
```
[the remark this answers is not in the document]
Comment (emolsson) [reply core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/RepairState.java]: I can add more description in a later patch

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/scheduling/ScheduledTask.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+     * Get the lock used by this scheduled job.
+     *
+     * @param lockFactory
+     *            The lock factory to use.
+     * @return The lock used by this scheduled job.
+     * @throws LockException Thrown when it's not possible to get the lock.
+     */
+    public LockFactory.DistributedLock getLock(LockFactory lockFactory) throws LockException
+    {
+        return lockFactory.tryLock(null, "SCHEDULE_LOCK", myPriority, new HashMap<>());
```
[the remark this answers is not in the document]
Comment (emolsson) [reply core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/scheduling/ScheduledTask.java]: Added DEFAULT_SCHEDULE_RESOURCE

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/RepairGroup.java [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+     * @param priority
+     *            The priority this lock has
+     * @return The acquired lock for the data center
+     * @throws LockException
+     *             Thrown when unable to get a lock in the data center
+     */
+    private LockFactory.DistributedLock getLockForDatacenter(LockFactory lockFactory, String dataCenter, int priority) throws LockException
+    {
+        Map<String, String> metadata = new HashMap<>();
+        metadata.put("keyspace", myTableReference.getKeyspace());
```
[the remark this answers is not in the document]
Comment (emolsson) [reply core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/RepairGroup.java]: Done

File: core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/RepairGroup.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (itskarlsson) [reply core/src/main/java/com/ericsson/bss/cassandra/ecchronos/core/repair/RepairGroup.java]: Sure

**Pull request merged**